### ISSUE TITLE
[Perf -31%] PerfLabTests.EnumPerf (2)

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/runtime#37805]
### Run Information

Architecture | x64
-- | --
OS | Windows 10.0.18362
Changes | [diff](https://github.com/dotnet/installer/compare/a4a49f07fbf986d496ef57521e76b22caaf9bfb3...ccd701fe23f236755650af3ce3d9ab6bdd8a4af9)

### Regressions in PerfLabTests.EnumPerf

Benchmark | Baseline | Test | Test/Base | Modality | Baseline Outlier
-- | -- | -- | -- | -- | --
ObjectGetType | 1.25 ns | 1.67 ns | 1.33 |  | False
EnumCompareTo | 12.54 ns | 16.17 ns | 1.29 |  | False

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/06_12_2020/release/3.1.1xx%20-%3E%20master_x64_Windows%2010.0.18362/PerfLabTests.EnumPerf_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/06_12_2020/release/3.1.1xx%20-%3E%20master_x64_Windows%2010.0.18362/PerfLabTests.EnumPerf.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'PerfLabTests.EnumPerf*';
```
<details>

#### PerfLabTests.EnumPerf.ObjectGetType

```log
[0.975 ; 1.115) | @@@
[1.115 ; 1.178) | 
[1.178 ; 1.318) | @@@@@@@@@
[1.318 ; 1.495) | @@@@@@@@@@@@@@@@@@@@@@@@@
[1.495 ; 1.649) | @@@@
[1.649 ; 1.822) | @@@@@@@@@@@@@@@@@
[1.822 ; 1.962) | @
```
#### PerfLabTests.EnumPerf.EnumCompareTo(color: Red)

```log
[12.189 ; 13.518) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[13.518 ; 14.907) | @@@@
[14.907 ; 15.979) | 
[15.979 ; 17.308) | @@@@@@@@@@@@@@@@@
[17.308 ; 18.845) | @@
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by DrewScoggins** [comment]
@jkotas @adamsitnik

**Comment by danmosemsft** [comment]
No helpful diff here. Need to repro locally comparing 3.1 to 5.0.

**Comment by danmosemsft** [comment]
@DrewScoggins @adamsitnik the repro command above does not work (at least on Windows, which is what this issue is for):
```
C:\git>py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'PerfLabTests.EnumPerf*'
...
[2020/08/02 12:25:35][INFO] $ pushd "C:\git\performance\src\benchmarks\micro"
[2020/08/02 12:25:35][INFO] $ dotnet run --project C:\git\performance\src\benchmarks\micro\MicroBenchmarks.csproj --configuration Release --framework netcoreapp5.0 --no-restore --no-build -- --filter 'PerfLabTests.EnumPerf*' --packages C:\git\performance\artifacts\packages --runtimes netcoreapp5.0
[2020/08/02 12:25:38][INFO] You must have made a typo in ''PerfLabTests.EnumPerf*''. Suggestions:
[2020/08/02 12:25:38][INFO]     PerfLabTests.EnumPerf
```
Changing to double quotes, and removing the semicolon, seems to be necessary, ie
```
C:\git>py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter "PerfLabTests.EnumPerf*"
...
[2020/08/02 12:27:47][INFO] $ pushd "C:\git\performance\src\benchmarks\micro"
[2020/08/02 12:27:47][INFO] $ dotnet run --project C:\git\performance\src\benchmarks\micro\MicroBenchmarks.csproj --configuration Release --framework netcoreapp5.0 --no-restore --no-build -- --filter PerfLabTests.EnumPerf* --packages C:\git\performance\artifacts\packages --runtimes netcoreapp5.0
[2020/08/02 12:27:50][INFO] // Validating benchmarks:
```
Could you please correct the script you use to generate these issues? (Perhaps single quotes/semicolon is correct for Unix)

**Comment by danmosemsft** [comment]
Running tests on my machine (benchmarks_ci.py makes this super easy!):
```
[2020/08/02 12:28:58][INFO]
[2020/08/02 12:28:58][INFO] BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.421 (2004/?/20H1)
[2020/08/02 12:28:58][INFO] Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
[2020/08/02 12:28:58][INFO] .NET Core SDK=5.0.100-rc.1.20401.3
[2020/08/02 12:28:58][INFO]   [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.40101, CoreFX 5.0.20.40101), X64 RyuJIT
[2020/08/02 12:28:58][INFO]   Job-QAXEFP : .NET Core 5.0.0 (CoreCLR 5.0.20.40101, CoreFX 5.0.20.40101), X64 RyuJIT
[2020/08/02 12:28:58][INFO]
[2020/08/02 12:28:58][INFO] PowerPlanMode=00000000-0000-0000-0000-000000000000  Runtime=.NET Core 5.0  Arguments=/p:DebugType=portable
[2020/08/02 12:28:58][INFO] Toolchain=netcoreapp5.0  IterationTime=250.0000 ms  MaxIterationCount=20
[2020/08/02 12:28:58][INFO] MinIterationCount=15  WarmupCount=1
[2020/08/02 12:28:58][INFO]
[2020/08/02 12:28:58][INFO] |                Method | color |       Mean |     Error |    StdDev |     Median |        Min |        Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/02 12:28:58][INFO] |---------------------- |------ |-----------:|----------:|----------:|-----------:|-----------:|-----------:|-------:|------:|------:|----------:|
[2020/08/02 12:28:58][INFO] |         ObjectGetType |     ? |  0.9879 ns | 0.0313 ns | 0.0261 ns |  0.9765 ns |  0.9531 ns |  1.0481 ns |      - |     - |     - |         - |
[2020/08/02 12:28:58][INFO] | ObjectGetTypeNoBoxing |     ? |  0.8632 ns | 0.0388 ns | 0.0363 ns |  0.8530 ns |  0.7976 ns |  0.9424 ns |      - |     - |     - |         - |
[2020/08/02 12:28:58][INFO] |            EnumEquals |     ? |  8.4196 ns | 0.1049 ns | 0.0930 ns |  8.4072 ns |  8.2411 ns |  8.5820 ns | 0.0115 |     - |     - |      48 B |
[2020/08/02 12:28:58][INFO] |         EnumCompareTo |   Red | 10.9006 ns | 0.1094 ns | 0.0914 ns | 10.9084 ns | 10.7876 ns | 11.1105 ns | 0.0114 |     - |     - |      48 B |

[2020/08/02 12:31:14][INFO]
[2020/08/02 12:31:14][INFO] |                Method | color |       Mean |     Error |    StdDev |     Median |        Min |        Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/02 12:31:14][INFO] |---------------------- |------ |-----------:|----------:|----------:|-----------:|-----------:|-----------:|-------:|------:|------:|----------:|
[2020/08/02 12:31:14][INFO] |         ObjectGetType |     ? |  0.9608 ns | 0.0271 ns | 0.0253 ns |  0.9601 ns |  0.9083 ns |  1.0002 ns |      - |     - |     - |         - |
[2020/08/02 12:31:14][INFO] | ObjectGetTypeNoBoxing |     ? |  0.8484 ns | 0.0185 ns | 0.0164 ns |  0.8526 ns |  0.8140 ns |  0.8725 ns |      - |     - |     - |         - |
[2020/08/02 12:31:14][INFO] |            EnumEquals |     ? |  8.4416 ns | 0.0867 ns | 0.0768 ns |  8.4241 ns |  8.3406 ns |  8.5934 ns | 0.0115 |     - |     - |      48 B |
[2020/08/02 12:31:14][INFO] |         EnumCompareTo |   Red | 11.1579 ns | 0.0864 ns | 0.0809 ns | 11.1470 ns | 11.0399 ns | 11.2999 ns | 0.0115 |     - |     - |      48 B |

[2020/08/02 12:33:20][INFO]
[2020/08/02 12:33:20][INFO] BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.421 (2004/?/20H1)
[2020/08/02 12:33:20][INFO] Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
[2020/08/02 12:33:20][INFO] .NET Core SDK=5.0.100-rc.1.20401.3
[2020/08/02 12:33:20][INFO]   [Host]     : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT
[2020/08/02 12:33:20][INFO]   Job-TULVLT : .NET Core 3.1.6 (CoreCLR 4.700.20.26901, CoreFX 4.700.20.31603), X64 RyuJIT
[2020/08/02 12:33:20][INFO]
[2020/08/02 12:33:20][INFO] PowerPlanMode=00000000-0000-0000-0000-000000000000  Runtime=.NET Core 3.1  Arguments=/p:DebugType=portable
[2020/08/02 12:33:20][INFO] Toolchain=netcoreapp3.1  IterationTime=250.0000 ms  MaxIterationCount=20
[2020/08/02 12:33:20][INFO] MinIterationCount=15  WarmupCount=1
[2020/08/02 12:33:20][INFO]
[2020/08/02 12:33:20][INFO] |                Method | color |       Mean |     Error |    StdDev |     Median |        Min |        Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/02 12:33:20][INFO] |---------------------- |------ |-----------:|----------:|----------:|-----------:|-----------:|-----------:|-------:|------:|------:|----------:|
[2020/08/02 12:33:20][INFO] |         ObjectGetType |     ? |  0.6554 ns | 0.0242 ns | 0.0214 ns |  0.6550 ns |  0.6241 ns |  0.6938 ns |      - |     - |     - |         - |
[2020/08/02 12:33:20][INFO] | ObjectGetTypeNoBoxing |     ? |  0.7557 ns | 0.0675 ns | 0.0599 ns |  0.7396 ns |  0.6737 ns |  0.8946 ns |      - |     - |     - |         - |
[2020/08/02 12:33:20][INFO] |            EnumEquals |     ? |  8.4566 ns | 0.1107 ns | 0.0981 ns |  8.4744 ns |  8.3060 ns |  8.6393 ns | 0.0115 |     - |     - |      48 B |
[2020/08/02 12:33:20][INFO] |         EnumCompareTo |   Red | 11.3875 ns | 0.0888 ns | 0.0831 ns | 11.3963 ns | 11.1871 ns | 11.5313 ns | 0.0115 |     - |     - |      48 B |

[2020/08/02 12:35:02][INFO]
[2020/08/02 12:35:02][INFO] |                Method | color |      Mean |     Error |    StdDev |    Median |       Min |       Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/02 12:35:02][INFO] |---------------------- |------ |----------:|----------:|----------:|----------:|----------:|----------:|-------:|------:|------:|----------:|
[2020/08/02 12:35:02][INFO] |         ObjectGetType |     ? |  1.055 ns | 0.0374 ns | 0.0313 ns |  1.057 ns |  1.008 ns |  1.118 ns |      - |     - |     - |         - |
[2020/08/02 12:35:02][INFO] | ObjectGetTypeNoBoxing |     ? |  1.089 ns | 0.0237 ns | 0.0210 ns |  1.088 ns |  1.059 ns |  1.123 ns |      - |     - |     - |         - |
[2020/08/02 12:35:02][INFO] |            EnumEquals |     ? |  8.322 ns | 0.0965 ns | 0.0903 ns |  8.327 ns |  8.138 ns |  8.461 ns | 0.0115 |     - |     - |      48 B |
[2020/08/02 12:35:02][INFO] |         EnumCompareTo |   Red | 11.436 ns | 0.1129 ns | 0.1056 ns | 11.431 ns | 11.274 ns | 11.610 ns | 0.0114 |     - |     - |      48 B |

[2020/08/02 12:39:06][INFO] |                Method | color |       Mean |     Error |    StdDev |     Median |        Min |        Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/02 12:39:06][INFO] |---------------------- |------ |-----------:|----------:|----------:|-----------:|-----------:|-----------:|-------:|------:|------:|----------:|
[2020/08/02 12:39:06][INFO] |         ObjectGetType |     ? |  0.6412 ns | 0.0420 ns | 0.0351 ns |  0.6470 ns |  0.5830 ns |  0.6985 ns |      - |     - |     - |         - |
[2020/08/02 12:39:06][INFO] | ObjectGetTypeNoBoxing |     ? |  0.7297 ns | 0.0416 ns | 0.0390 ns |  0.7397 ns |  0.6350 ns |  0.7686 ns |      - |     - |     - |         - |
[2020/08/02 12:39:06][INFO] |            EnumEquals |     ? |  8.4391 ns | 0.1033 ns | 0.0915 ns |  8.4460 ns |  8.2638 ns |  8.5789 ns | 0.0114 |     - |     - |      48 B |
[2020/08/02 12:39:06][INFO] |         EnumCompareTo |   Red | 11.3681 ns | 0.1423 ns | 0.1331 ns | 11.3452 ns | 11.1288 ns | 11.6325 ns | 0.0114 |     - |     - |      48 B |
```

EnumCompareTo has not regressed: it's gotten a little faster. ObjectGetType and ObjectGetTypeNoBoxing seem to have regressed a little. 

I believe Object.GetType() is code owned by the VM. Moving there.

**Comment by jkotas** [comment]
This was originally filled as https://github.com/DrewScoggins/performance-2/issues/525 and fixed by #37845
